### ISSUE TITLE
Reduce code duplication for javascript

### ISSUE THREAD

**Issue by gerzytet** [wysohn/TriggerReactor#198]
Nearly all executors and placeholders have some boilerplate code to length-check the argument list.
This is an obvious source of duplication that could be eliminated
My suggested solution to this is to add some metadata to the executors and placeholders, probably in the form of:
```
minArgs = 1
maxArgs = 3
function EXAMPLE () {
/...
}
```

These values will be extracted when the js gets loaded, and used to validate the length.

Benefits:
- Standardized error messages as a result of this will make writing tests easier
- Shorted and less redundant js code

**Comment by wysohn** [comment]
Sounds good to me!

**Comment by wysohn** [comment]
Though, what would be the expected js code if this is implemented?

Do you have anything in mind?

What if the minArgs is 1 and maxArgs is 3, yet the Executor or Placeholder does not support 2 arguments?

**Comment by gerzytet** [comment]
Hmm.  Maybe it should be something like
SupportedArgs = {1, 3} then

**Comment by rkdrns4747** [comment]
@gerzytet can you give me some more example?

**Comment by gerzytet** [comment]
@rkdrns4747 
Many executors can be called with multiple types of args.
Let's say that #KICK could be called in the following ways:
```
#KICK 
#KICK "gerzytet"
#KICK "gerzytet", "destroying spawn"
```
#KICK would have the following validation object:
```
validation = {
	"overloads" : [
		[
			
		],
		[
			{
				"type" : "string"
			}
		],
		[
			{
				"type" : "string"
			},
			{
				"type" : "string"
			}
		]
	]
}
```
overloads is a list of all the ways the executor can be used.  It contains 3 items.
The first one is empty.  This represents the case where the executor is called with no args.
The second one has a single arg with the type set to string.
The third one has 2 args with a type of string.

When the executor is called, the validation will look through this list of overloads in order and try to find one that has the same length and matches all the criteria.

If it does, it sets a variable called "overload" which represents which overload was chosen.
In this case, calling #KICK with no args would set it to 0, since it matches the first overload.

The next thing I'm going to do with this is to implement error-message generation.  This will require an extra "name" attribute so it can generate things like "argument x must be a number"

**Comment by wysohn** [comment]
@gerzytet 
Looks really good.

Wouldn't it also be good to have a format like below?

    validation = {
        "overloads" : [
            [],
            [{"type" : "string"}],
            [{"type" : "string"}, {"type" : "string"}]
        ]
    }

It's easier to see each overload predicate nice and clean.

Or, is there any possible way to make it much simpler?

    validation = [
        [],
        ["string"],
        ["string", "string"]
    ]

The above code might be an easier way (and also intuitive) if such syntax is possible.

**Comment by gerzytet** [comment]
@wysohn 
In terms of formatting, you can format it however you want.  js parses it, not TR

Even though I haven't demonstrated this yet, the validation can take properties other than "type"
```
validation = {
    overloads = [
        [{"type" : "int", "minimum" : 0}]
    ]
}
```
Currently, only type and minimum are supported, but I plan to have more.
The one I'm working on next is a "name" attribute to help generate error messages
```
{"type" : "string", "name" : "world"}
```
might result in an error like `world must be a string. got: 5`

The validation also doesn't care what order you put the overloads in.   you can have 2 args, then 0 args, then 3 args if you want to

**Comment by gerzytet** [comment]
@wysohn 
Because of this, inserting null isn’t necessary or valid

It’s true that we could have validation directly be a list like:
```
validation = [
        [],
        [{"type" : "string"}],
        [{"type" : "string"}, {"type" : "string"}]
]
```

But I also plan to eventually have top-level properties for the validation, which wouldn’t work if validation was an array:
```
validation = {
    "deprecated": true
    "overloads": [...]
}
```
A "deprecated" property would allow us to get rid of the hard-coded list in ExecutorManager.

One more thing to mention: if you want an arg that accepts any value at all, you can use empty braces {}
This will accept 2 args of any type (or at least it should, haven’t tested):
```
[{}, {}]
```
This works because every option you add to an arg acts as a filter.  "type" is a filter for the type.  If you remove the type filter, it won’t do any validation on the type.

**Comment by wysohn** [comment]
@gerzytet 
Brilliant!

Just to add one more thing; are we really going to in need other than "type" and "name"?

Thinking in terms of functions in most common languages, all we need to know is the type and the name of the input variable. So I wonder if there will be any case other than providing the type of and name of the variable.

If it's not, we can reduce it a little bit more I suppose.

    validation = {
        "deprecated": false,
        "overloads": [
            null,
            null,
            null,
            null,
            [{"string":"world"}, {"number":"x"}, {"number":"y"}, {"number":"z"}]
        ]
    }

I see you have "minimum" as another option, yet I think we can write the 'predicate' of argument in the actual js code.

Think about an example where we can accept multiple overloaded arguments. In the way how it handles the argument now, we will have to define "minimum" for every single arguments.

    validation = {
        "deprecated": false,
        "overloads": [
            null,
            null,
            [{"type":"string", "name":"world"}, {"type":"number", "name":"x", "minimum": 5}],
            [{"type":"string", "name":"world"}, {"type":"number", "name":"x", "minimum": 5}, {"type":"number", "name":"y"}],
            [{"type":"string", "name":"world"}, {"type":"number", "name":"x", "minimum": 5}, {"type":"number", "name":"y"}, {"type":"number", "name":"z"}]
        ]
    }

It will be too tedious in my opinion.

I think it would be better to leave it in js to handle is much easier since

    var SOMEEXECUTOR(args){
        if(args[1] < 5)
            throw new Error("x coordinate has to be larger or equal to 5!")
    }

---

And for the null, I'm using it so that the actual array will fit into the index where it should be.

For the above code for example, since we want to accept only the argument with three variables (world, x, y, z), it has to be in the index of 3, so it can match up with the number of input arguments.

I'm keep suggesting the 'null' way of argument setting because I find it more intuitive, in my opinion, and it allows O(1) access time, but I will leave it as your choice. Technically speaking, since the "overloads" is the array of N elements, if the Executor is put in some kind of N times loop, the complexity will be O(N^2) which is not really desirable. If, however, the access time of "overloads" is guaranteed to be constant, it will be always O(N).

Plus, can we use "arguments" instead of "overloads"? Even the name of the variable is 'args' in the actual Executor and Placeholder code, so it would make much more sense that way.

**Comment by gerzytet** [comment]
@wysohn 
*And why so many statics!*
There aren't many places in the code where I used static when I shouldn't have.  Please be more specific

There's a reason I decided to build the validation interface the way I did
Your suggestion:
```
boolean validate(JsObject validationInfo, Object[] arguments){
    some mysterious things happen here...
}
```
Means that the validation data will be extracted from the JsObject *every time the placeholder is called*.  This is not a trivial process, and it has an entire validation process in itself.  To make it so that it would only be done once, the validation is instead done through a `Validator` Object.  You can make one from a JSObject:
```
public static Validator from(JSObject js)
```
then use it for validation:
```
public int validate(Object... args)
```
You also suggested an additional:
```
interface PostTask{
    void run(boolean validationResult);
}
```
I don't think we would ever need this.  If you can name a situation where it would reduce complexity and repetition that exists right now, I'll add it.

Your suggestion about this format:
```
validation = {
    "deprecated": false,
    "overloads": [
        null,
        null,
        null,
        null,
        [{"string":"world"}, {"number":"x"}, {"number":"y"}, {"number":"z"}]
    ]
}
```
Is not applicable for 2 reasons:
1. You mention performance concerns when trying to find the best overload based on length.  While it is true that forcing a consistent ordering for the args would allow O(1) access, it isn't a performance concern.  When you consider performance, you need to think about what the *real world* usage is.  Most executors have between 1-2 overloads.  I cant think of any that could have more than 4.  Additionally, finding out that an overload does not match based on length is the first thing that gets checked, and it's as simple as reading args.length.  If executors had 50 overloads, it would become a performance concern, but right now, it's trivial.
2. Overloads are built to handle overloads in any order at all, and to handle overloads that have the same length.  consider this example:
```
[{"type" : "int"}],
[{"type" : "string"}],
[{}]
```
Each overload is 1 arg long, but the validation can still distinguish based on the types and any other options supplied.  If overload length was its index, this wouldn't be possible.

If the length of validation objects bothers you, I suppose you can do this:
```
world = {"type":"string", "name":"world"}
x = {"type":"number", "name":"x", "minimum": 5}
y = {"type":"number", "name":"y"}
z = {"type":"number", "name":"z"}
validation = {
    "overloads": [
        [world, x],
        [world, x, y],
        [world, x, y, z]
    ]
}
```

**Comment by wysohn** [comment]
Pretty much everything is about how flexible the system is so that it can adapt to the future change with minimal impact on the project.

---

> *And why so many statics!*
> There aren't many places in the code where I used static when I shouldn't have.  Please be more specific

It's just my personal preference where static should be avoided when it can be done without it. I suppose the static{} initialization field part can be done in the actual Map instance instead of doing it in each of the child classes. https://github.com/wysohn/TriggerReactor/blob/461e0cde7c870317778b4ca9d77bd7f943cc4e35/core/src/main/java/io/github/wysohn/triggerreactor/core/script/validation/option/TypeOption.java#L7

However! it would be micromanagement if I tell you to do so; I will leave it as your choice. I used to have this pattern in my old codes too. It doesn't violate anything.

---

> There's a reason I decided to build the validation interface the way I did
> Your suggestion:
> ```
> boolean validate(JsObject validationInfo, Object[] arguments){
>     some mysterious things happen here...
> }
> ```
> Means that the validation data will be extracted from the JsObject *every time the placeholder is called*.  This is not a trivial process, and it has an entire validation process in itself.  To make it so that it would only be done once, the validation is instead done through a `Validator` Object.  You can make one from a JSObject:
> ```
> public static Validator from(JSObject js)
> ```
> then use it for validation:
> ```
> public int validate(Object... args)
> ```
> You also suggested an additional:
> ```
> interface PostTask{
>     void run(boolean validationResult);
> }
> ```
> I don't think we would ever need this.  If you can name a situation where it would reduce complexity and repetition that exists right now, I'll add it.

The suggestion is just an example. What I'm saying is that you don't have to put everything in one basket when planning on a system. 

```
boolean validate(Validator validationInfo, Object[] arguments){
    some mysterious things happen here...
}
```

This will be just identical to what you have said. I'm not saying that you have to change something; I'm just saying that dividing the tasks into more abstract layers allows us to think more intuitively, thus, fewer errors we will make.

We started with this idea

```
boolean validate(JsObject validationInfo, Object[] arguments){
    some mysterious things happen here...
}
```

And realized that this can be divided up to two modules

```
public static Validator from(JSObject js){
}

//Save and reuse the Validator instance

boolean validate(Validator validationInfo, Object[] arguments){
    some mysterious things happen here...
}
```

And further, we realize that we have to provide feedback to the user

```

interface PostTask{
    void run(String whatWentWrong);
}

public static Validator from(JSObject js){
}

//Save and reuse the Validator instance

void validate(Validator validationInfo, Object[] arguments, PostTask post){
    if something not passed the test...
        post.run(validationInfo+" failed?") 
}

someValidation.validate(savedValidator, arguments, (cause)->{print cause?})
```

Anyway, you didn't need the advice since you already knew about it. 

---

> Your suggestion about this format:
> ```
> validation = {
>     "deprecated": false,
>     "overloads": [
>         null,
>         null,
>         null,
>         null,
>         [{"string":"world"}, {"number":"x"}, {"number":"y"}, {"number":"z"}]
>     ]
> }
> ```
> Is not applicable for 2 reasons:
> 1. You mention performance concerns when trying to find the best overload based on length.  While it is true that forcing a consistent ordering for the args would allow O(1) access, it isn't a performance concern.  When you consider performance, you need to think about what the *real world* usage is.  Most executors have between 1-2 overloads.  I cant think of any that could have more than 4.  Additionally, finding out that an overload does not match based on length is the first thing that gets checked, and it's as simple as reading args.length.  If executors had 50 overloads, it would become a performance concern, but right now, it's trivial.

Big O analysis is nothing more than measuring the worst-case scenario. That's why I'm saying that the argument search will take N times even we can see that the number of arguments will be less than 10 most of the time. Most importantly, the argument search will take place in the non-server thread, so it would never cause any performance issue to the server. 

However, if we have to change the design in the future because of the performance issue, it will cost us more than do it right now since we are going to write more codes based on this design. So why risk it when there is an easier way?

Also, please read the comment below to see my idea behind it. This and the below are related.

> 2. Overloads are built to handle overloads in any order at all, and to handle overloads that have the same length.  consider this example:
> ```
> [{"type" : "int"}],
> [{"type" : "string"}],
> [{}]
> ```
> Each overload is 1 arg long, but the validation can still distinguish based on the types and any other options supplied.  If overload length was its index, this wouldn't be possible.
> 
> If the length of validation objects bothers you, I suppose you can do this:
> ```
> world = {"type":"string", "name":"world"}
> x = {"type":"number", "name":"x", "minimum": 5}
> y = {"type":"number", "name":"y"}
> z = {"type":"number", "name":"z"}
> validation = {
>     "overloads": [
>         [world, x],
>         [world, x, y],
>         [world, x, y, z]
>     ]
> }
> ```

Yes, I actually thought about the case too, but do we need that complex argument? Executors usually have one simple job, so I question the necessity of the amount of complexity just for the arguments. It seems like an overshoot for me. Some .js code will have more codes in argument validation than the actual body of the code. 

> [{"type" : "int"}],
> [{"type" : "string"}],

For the case where we need the argument that has the same number of arguments with different types, we can just leave it as {}, so it will not be checked, then we can manually do it in the .js code. 

I already can see the number of hassles you have to deal with in order to find the appropriate validator which will match with the given argument.

What if the overloads are like this?

> [{"type" : "int"}, {"type":"number"}],
> [{"type" : "number"}, {"type":"int"}],

It's very ambiguous situation, and checking the validity of validation itself will be very hard too.

**Comment by wysohn** [comment]
@gerzytet 
bump

**Comment by gerzytet** [comment]
@wysohn 
> It's just my personal preference where static should be avoided when it can be done without it. I suppose the static{} initialization field part can be done in the actual Map instance instead of doing it in each of the child classes.

I was trying to make it so you could add new ValidationOptions without needing to change ValidationOption, and so that most classes wouldn't need to reference option classes directly.
I ended up failing at the first one (classes have to be interacted with in some way before their static blocks are run), but succeeded at the second.
This can be refactored so that it removes register, since it isn't possible in the way I imagined

> Big O analysis is nothing more than measuring the worst-case scenario. That's why I'm saying that the argument search will take N times even we can see that the number of arguments will be less than 10 most of the time. Most importantly, the argument search will take place in the non-server thread, so it would never cause any performance issue to the server.

If the performance is an issue,  there are much bigger targets to knock down.  Consider that  *every time $x is called, it has to boot up a javascript engine*.  That's the biggest problem of all.  Also, when a method is called, the interpreter thinks that it has never seen it before and uses reflection to look up the method all over again.
I can't prove that either of these things are causing performance issues (although I'm pretty sure they are), but you have even less proof since the only working example of overloads in the whole code base has only 2.  You can't know if there's a performance issue or not until you use it!

If you want to crack down on performance, there are more worthwhile things to do.

Even if it becomes a problem in the future, we can fix it at no cost by simply re-ordering the overloads as they are loaded.  No validation info has to be changed.
That's the real power of the validation system.  It's data, not code, so you can transform it in more interesting ways.


There are plenty of Executors that need overloads of the same length.  consider:
- Executors that accept players and player names
- Executors that accept items and item names
- Executors that accept locations and xyz
- #MODIFYPLAYER, #SCOREBOARD, and #MODIFYHELDITEM

Once we directly support this, more executors can be added to use same-size overloads in creative and useful ways.

Also, unlike in Java, *it's impossible to create an ambiguous overload list*, even if I were to write something stupid like this:
```
[{"type" : "number"}],
[{"type" : "number"}]
```

The validation follows a simple procedure:
1. look at the first overload
2. check if it matches
3. if it doesn't, pick the next one and repeat 2.

There's no room for ambiguity.  It either matches or it doesn't.  If it doesn't the next one is picked.
In the example above, the second overload will never be picked since the first one is considered first.

**Comment by wysohn** [comment]
@gerzytet 

> If the performance is an issue, there are much bigger targets to knock down. Consider that every time $x is called, it has to boot up a javascript engine. That's the biggest problem of all. Also, when a method is called, the interpreter thinks that it has never seen it before and uses reflection to look up the method all over again.

1. That's not relevant to this issue at all. We always have to strive for the best result. $x performing poorly doesn't do anything with this issue. You may open a new issue if that's problematic.

2. Even if it was relevant, what you have said is not true. You can check it yourself using the new timings system that was newly added. Executors and Placeholders do not boot up the javascript engine multiple times. Instead, after the first execution, it compiles the .js file so that the consecutive calls will be cached thus will run much faster. You can check it in the code yourself, or you may use the timings command to double-check that the average execution time of Executor or Placeholder is between 0~2ms.

> I can't prove that either of these things are causing performance issues (although I'm pretty sure they are), but you have even less proof since the only working example of overloads in the whole code base has only 2. You can't know if there's a performance issue or not until you use it!

I cannot perform any clairvoyance, so I have to rely on mathematical proofs like Big O analysis. I can prove that the execution time will be O(N^2) mathematically if you want to, even though I'm no expert at it. And O(N^2) is even just a very optimistic estimate where we only have a single Executor inside a while loop, and if it's inside of nested loop instead of just a single while loop, it would be even worse. Don't forget that we use multiple Executors and Placeholders. As I said, I do not possess any clairvoyance ability, so Big O analysis is pretty much I can do to make my best prediction. 

For my example, it was only to show a single point where it can be a problem. Your example

```
[{"type" : "number"}],
[{"type" : "number"}]
```

does not reflect what I'm concerned about. I'm talking about the ambiguity when there are parent and child relationship exists with the given argument.

My example again

> [{"type" : "int"}, {"type":"number"}],
> [{"type" : "number"}, {"type":"int"}],

In this case, we can expect some input of (integer, integer), (integer, any number), or (any number, integer), then how are you going to distinguish the ambiguity? if input is (1,1), which signature should we use? Again, this is just one example that can happen, and there are like infinite possible combinations where such ambiguity can arise.

You said that as a solution, we can just compare them one by one from top to bottom and discard the rest as soon as we find something that matches, then why do we need such a complex system? Can't we just use simple indexing and use {} where there can be possible ambiguity happens? I don't really see much benefit.

Of course, you can implement it just like how Java or other modern languages handle it, yet why would you put that much effort just for arguments? I had to do similar stuff for direct Java method invocation since I couldn't find any other way around, and it wasn't really fun thing to work on. But it's worthwhile since now TR can access any java methods directly so that we don't have to add support for different plugins whenever users require them.

**Comment by gerzytet** [comment]
@wysohn 
> Executors and Placeholders do not boot up the javascript engine multiple times. Instead, after the first execution, it compiles the .js file so that the consecutive calls will be cached thus will run much faster. You can check it in the code yourself, or you may use the timings command to double-check that the average execution time of Executor or Placeholder is between 0~2ms.

You've contradicted yourself.  Running js involves running (or "booting up") the engine.  Even though it's faster the second time, it's still a lot of overhead compared to calling a java method.

If you follow the process I laid out, you would know that the first overload is chosen for (1, 1).  The first overload matches and the first overload is considered first, so it is chosen.

> You said that as a solution, we can just compare them one by one from top to bottom and discard the rest as soon as we find something that matches
This is *already* how it works.

The goal of the validation system isn't to reduce the length of the code
This might happen anyway, but the main goal is to take advantage of benefits offered by the [rule of least power](https://en.wikipedia.org/wiki/Rule_of_least_power)

The most powerful thing about the validation system is that the validation isn't done in code, it's represented as data.  We can use this data in interesting ways, such as:
- error message generation for failed validations (what I'm working on now)
- unit testing validation without running the executor or handling exceptions
- static type checking for executor/placeholder inputs
- optimizations that can *skip* validation if the inputs are known
- documentation generation

If it's validated in code, we can't do much with it.  Only run it.  For all of the above to work, we have to have to write all the validation as data, even if it makes the code longer.


If you still believe that overload matching is causing issues, wait a couple days until I'm finished with error message generation, then change it.  I personally think changing it is a waste of time.  O(N ^ 2) is only a problem if N is big... N is usually around 2

**Comment by wysohn** [comment]
> You've contradicted yourself. Running js involves running (or "booting up") the engine. Even though it's faster the second time, it's still a lot of overhead compared to calling a java method.

Well, to be honest, I do not know the exact meaning of "booting up" you used here. So let's put that aside since it's not that necessary. And of course, using a script engine is slower because it has to pass through two layers (both jvm and js engine) in order to work. If someone wants that much performance boost, he or she will have to write their own plugin. (Or, they can register Executor manually which doesn't involve Javascript. If you noticed, #CMDOP is written purely in Java for security reason, and it also provides better performance)

> If you follow the process I laid out, you would know that the first overload is chosen for (1, 1). The first overload matches and the first overload is considered first, so it is chosen.

Yes, I already got what you have said. Top to bottom, check one by one.

> The goal of the validation system isn't to reduce the length of the code
> This might happen anyway, but the main goal is to take advantage of benefits offered by the rule of least power
> 
> The most powerful thing about the validation system is that the validation isn't done in code, it's represented as data. We can use this data in interesting ways, such as:

Yes, I am with you about this idea from the beginning. My only concern is why do we need such a complex system when it's not likely required.

In fact, if I read it correctly, the 'rule of least power' should give more flexibility to the language. Defining a strict function signature is, in fact, making it less flexible.

> error message generation for failed validations (what I'm working on now)

We can still generate a message for invalid number of arguments provided or for the type of given argument being different. Which I think brilliant.

> unit testing validation without running the executor or handling exceptions

True, but unit testing the validation itself is nothing to do with the Executor. Unit testing on argument validation only guarantees that the domain of input will be as defined, yet Executor still can malfunction with the correct inputs provided, thus, we still need unit testing for the Executors separately.

> static type checking for executor/placeholder inputs

We can still make static type check with index-based arguments, with the exception of bypassing the type check if the type can be ambiguous.

> optimizations that can skip validation if the inputs are known

Not sure what it means

> documentation generation

Like how? by generating Javadocs?

> If it's validated in code, we can't do much with it. Only run it. For all of the above to work, we have to have to write all the validation as data, even if it makes the code longer.

I'm not saying that we should get rid of all the validation systems we are discussing about here. I'm saying that it shouldn't be too complex. Why not just check number of arguments and the type? Either using the design you proposed or the design I suggested, we still have to check the length of `args` in the .js code, since it can be multiple.

Imagine where we have predicates:

```
[{"type" : "int"}],
[{"type" : "int"}, {"type":"string"}],
[{"type" : "string"}, {"type":"int"}],
```

if the input is ("myname",1), inside the .js, we will need some code like this:

```
var MYEXECUTOR(args){
    if(args.length == 1){
        //do something for (int)
    } else if(args.length == 2){
        var someInt = args[0];
        var someStr = args[1];
    }
}
```

Notice that this makes a very ambiguous situation because those variables with arg length 2 now have two possible values that are not even the same type.

So in the end, even though the validator checked on the type, we still have to check the actual type of args[0] and args[1] in order to distinguish which signature we have to deal with. So why don't we just leave it as {} so it can be handled internally?

If we have only one 'args' signature for one argument length (with indexing), we do not have to worry about checking the value, since we are certain that what will be the type of args[0] or args[1]. And, if we have set it {} so the args[0] or args[1] is not known, we can just internally check the type manually inside the .js code.


> If you still believe that overload matching is causing issues, wait a couple days until I'm finished with error message generation, then change it. I personally think changing it is a waste of time. O(N ^ 2) is only a problem if N is big... N is usually around 2

This is actually less concern than the design itself. It's something that is additionally caused by this design, but my main concern is that the system is just too complicated when it doesn't have to be. We can imagine how much work can be saved if the size of `args` and the types in args are matching, yet in my opinion, we do not have to perform 'super detailed' validation of every argument since we will have to check them in the .js once again.

Also, because the Executor and Placeholder run one by one in server thread 'per one server tick,' each execution will take at least 50ms anyway regardless of how efficient the script is. But fewer ms is better since if the server thread takes more than 50ms to process everything in one tick, the TPS will drop, and the server will lag (single thread intensifies).

**Comment by wysohn** [comment]
@gerzytet 
bump

**Comment by gerzytet** [comment]
@wysohn 
by "booting up" I meant starting and running the engine.  I wasn't clear about that, sorry.

>'rule of least power' should give more flexibility to the language. Defining a strict function signature is, in fact, making it less flexible.

The idea is to make it less flexible so you use the information contained in the code in more ways.

>optimizations that can skip validation if the inputs are known
>Not sure what it means

What I mean is that we can skip validation for things like $random:1:10 since we know it's correct.

>documentation generation
>Like how? by generating Javadocs?

pretty much.  I've always liked the idea of in-game documentation for the executors/placeholders.

```
[{"type" : "int"}],
[{"type" : "int"}, {"type":"string"}],
[{"type" : "string"}, {"type":"int"}],
if the input is ("myname",1), inside the .js, we will need some code like this:

var MYEXECUTOR(args){
    if(args.length == 1){
        //do something for (int)
    } else if(args.length == 2){
        var someInt = args[0];
        var someStr = args[1];
    }
}
```

The validation system sets a variable inside the engine called `overload`, indicating which overload was chosen.  This is demonstrated inside the CLEARPOTION code:
```
if(player === null)
		return null;

	if(overload == 0){
		var activeEffects = player.getActivePotionEffects();
		for(var iter = activeEffects.iterator(); iter.hasNext();){
			var type = iter.next().getType();
			player.removePotionEffect(type);
		}
	}else{
		var typeName = args[0].toUpperCase();
		var PotionEffectType = Java.type('org.bukkit.potion.PotionEffectType');
		var type = PotionEffectType.getByName(typeName);
		
		if(type == null)
			throw new Error("Invalid PotionEffectType named "+typeName);
			
		player.removePotionEffect(type);
	}
```
No other checking needs to be done even in seemingly ambiguous situations, because you can just look at the overload variable.  Just to be clear: *executors written using this system should not check args.length*.

I believe the overload variable addresses most of your concerns

**Comment by wysohn** [comment]
@gerzytet 
That makes more sense now. Thanks!

Now I can see where you are heading to.

---

And also, it would be nice to have the 'fundamental variable checking' if you are not working on it yet.

Like #MESSAGE executor, for example, users often use it in the context where the 'player' variable cannot exist.

Using #MESSAGE in Repeating Trigger is a possible case. No player has trigger the Repeating Trigger, so how can it have 'player'? It just doesn't make sense, yet current system just spit out some ambiguous message like 'player' is not defined or it is null.

This will be also useful for #163, yet I'm not sure which way we can make it informative enough to easily noticeable by the user. 

Just an idea.

**Comment by gerzytet** [comment]
@wysohn 
In a lot of JavaScript, (especially placeholders!) non-existent players are handled by silently returning null.
It might not be the best way, but we should make this the official behavior for missing players, at least for placeholders.  It’s either that or another deprecation campaign.

As for all the other variables, we can add validation for those, and I have considered its usefulness for #163.  Compared to vt and skript, we lack in this area.
Still in progress on the errors though.  Most of this week was spent studying for exams and doing other stuff.  This weekend I’m out of town.

**Comment by wysohn** [comment]
@gerzytet 
Always focus on study first and then work on this stuff. I'm busy with studying too

**Comment by wysohn** [comment]
Hey, I just saw the commit. Everything looks great! Can't wait to migrate .js codes into the new format.

Thought for this line: https://github.com/wysohn/TriggerReactor/blob/eb5a270247f9630afc65d4808f3d9fd0e1ef90d8/core/src/main/java/io/github/wysohn/triggerreactor/core/manager/AbstractExecutorManager.java#L175

Maybe we can do that in the constructor instead of in execute() instead of using the boolean variable to check it?

The js codes will be compiled in the constructor, at the last two lines of the constructor, so there would be some way to retrieve the 'validation' JsObject using the compiled js code.

Just a minor concern though.

**Comment by gerzytet** [comment]
@wysohn 
This would mean that every executor would be run when the plugin loads.  That’s a pretty significant performance impact when the server starts.

**Comment by wysohn** [comment]
@gerzytet 
And that's what we want to do usually.

We don't really care much about how long it takes to the server to start, but the performance during the server runtime is actual issue.

One additional benefit by doing so as I proposed is that it will also warm up the compiled js codes, so by the time when the server is running, compiled code can run much faster. If you have not noticed, each Executor and Placeholder takes 100~200ms at first run and then 0~2ms from the second run because of this reason.

**Comment by gerzytet** [comment]
@wysohn 
If you want to implement it that way, the problem is deeper than the validation system itself.  For one thing, the executors construct and run the js with a new “bindings” object each time.
The firstRun field was the best way to prevent reloading the Validator each time while avoiding refactoring the whole class.

We need to start comíng up with ways to implement #163, I suggest a validation field that looks like this:
{
"context": "org.bukkit.entity.event....",
"overloads": [...]
}

Executor.execute has a parameter called e, which is an object.  The interpreter sets it to the event that triggered the executor.  All we need to do is validate e’s type to make an executor that only works with certain events.

**Comment by wysohn** [comment]
@gerzytet 
Though, the current implementation instantiates the JsExecutor every time when the /trg reload is issued, so even if we use the firstRun field, it will eventually get loaded again. Plus, we also have to expect that the Validator may change after /trg reload since it's part of the .js, and it can be in runtime.

**Comment by gerzytet** [comment]
@wysohn 
The average user will only use a fraction of the executors and placeholders though.  They don’t all need to be loaded.

**Comment by wysohn** [comment]
@gerzytet 
True, I think it's better to leave it as now.

**Comment by gerzytet** [comment]
@wysohn 
All the core features of the validation are done now.

**Comment by wysohn** [comment]
@gerzytet 
https://github.com/wysohn/TriggerReactor/wiki/Creating-Executor

This is as far as I understand. Correct me if I got something wrong.
